r/resource_provider_registration - support for managing `features`

ms-henglu opened the pull request:
The added tests are listed as the following.

```
=== RUN   TestAccResourceProviderRegistration_feature
=== PAUSE TestAccResourceProviderRegistration_feature
=== CONT  TestAccResourceProviderRegistration_feature
--- PASS: TestAccResourceProviderRegistration_feature (2980.70s)
PASS
```

jackofallops commented on azurerm/internal/services/resource/feature_resource_test.go:
```diff
@@ -0,0 +1,104 @@
+type FeatureResource struct {
+}
+
+func TestAccFeatureResource_basic(t *testing.T) {
+	data := acceptance.BuildTestData(t, "azurerm_feature", "test")
+	r := FeatureResource{}
+
+	data.DataSourceTest(t, []acceptance.TestStep{
+		{
+			Config: r.basic(),
+			Check: acceptance.ComposeTestCheckFunc(
+				check.That(data.ResourceName).ExistsInAzure(r),
+			),
+		},
+	})
+}
+
+func TestAccFeatureResource_requiresImport(t *testing.T) {
+	data := acceptance.BuildTestData(t, "azurerm_feature", "test")
+	r := FeatureResource{}
+	data.ResourceTest(t, r, []acceptance.TestStep{
+		{
+			Config: r.requiresImportBasic(),
+			Check: acceptance.ComposeTestCheckFunc(
+				check.That(data.ResourceName).ExistsInAzure(r),
+			),
+		},
+		data.RequiresImportErrorStep(r.requiresImport),
+	})
+}
```
Hi @ms-henglu - just noticed while passing through that these will need to be skipped (via `t.Skip()`) until out acceptance test setup can support having features toggled without potentially causing issue elsewhere in the testing suite.

ms-henglu replied:
Thanks! @jackofallops , I added skip on these tests.

jackofallops commented on azurerm/internal/services/resource/feature_resource_test.go:
```diff
@@ -0,0 +1,110 @@
+	return utils.Bool(true), nil
+}
+
+func (r FeatureResource) basic() string {
+	return `
+provider "azurerm" {
+  features {}
+}
+
+resource "azurerm_feature" "test" {
+  name               = "AutoApproveFeature"
+  provider_namespace = "Microsoft.CognitiveServices"
+}
+`
+}
+
```
If you can update this to: 
```suggestion
func (r FeatureResource) basic(data acceptance.TestData) string {
	return fmt.Sprintf(`
provider "azurerm" {
  features {}

  subscription_id = "%s"
}

resource "azurerm_feature" "test" {
  name               = "AutoApproveFeature"
  provider_namespace = "Microsoft.CognitiveServices"
}
`, data.Client().SubscriptionIDAlt)
}

```

and similarly for the other tests, we should be good to run these on the Acceptance Test environment.  I'll do a full review once that's in and I after figure out why the Unit Test GH Action is currently failing all PR's....

katbyte commented on azurerm/internal/services/resource/registration.go:
```diff
@@ -36,6 +36,7 @@ func (r Registration) SupportedDataSources() map[string]*pluginsdk.Resource {
 // SupportedResources returns the supported Resources supported by this Service
 func (r Registration) SupportedResources() map[string]*pluginsdk.Resource {
 	return map[string]*pluginsdk.Resource{
+		"azurerm_feature":                              resourceFeature(),
```
unless this isn't subscription onlt could we rename it to
```suggestion
		"azurerm_subscription_feature":                              resourceFeature(),
```

ms-henglu commented:
Hi @katbyte @jackofallops , thanks for the suggestions, I've added a commit according to them. Please take another look, thanks!

jackofallops commented on azurerm/internal/services/resource/feature_resource_test.go:
```diff
@@ -0,0 +1,111 @@
+}
+
+func TestAccFeatureResource_basic(t *testing.T) {
+	if true {
+		t.Skip("Skipping due to acceptance test doesn't support having features toggled currently")
+	}
```
This can be removed now the test env is ready in the alt subscription.

jackofallops commented on azurerm/internal/services/resource/feature_resource_test.go:
```diff
@@ -0,0 +1,111 @@
+}
+
+func TestAccFeatureResource_requiresImport(t *testing.T) {
+	if true {
+		t.Skip("Skipping due to acceptance test doesn't support having features toggled currently")
+	}
```
As above, this can be removed now the test env is ready in the alt subscription.

jackofallops commented on azurerm/internal/services/resource/feature_resource.go:
```diff
@@ -0,0 +1,198 @@
+	return nil
+}
+
+func featureRegisteringStateRefreshFunc(ctx context.Context, client *features.Client, id parse.FeatureId) pluginsdk.StateRefreshFunc {
```
Should this function have an error state for manually approved features to help users? (iirc these features have the state as "Pending" rather than "Registering"?)

ms-henglu replied:
Yes, I added that at line162, let user know manually approved features can't be managed by tf.

jackofallops reviewed (changes requested):
Hi @ms-henglu - just a couple minor changes so we can run the tests for this, and a question on the user experience for the refresh func for manually approved features (e.g. `Microsoft.HardwareSecurityModule/AzureDedicatedHSM`)

Additionally, do the providers need to be re-registered for terraform to pick up the features that are being activated to be useful in the same `apply`. If  we need to re-register Terraform's provider for users to have access to an activated feature, this may need integrating with `azurerm_resource_provider_registration`?

ms-henglu commented:
@jackofallops Hi, thanks for code review.

I've removed skip on these tests.

>just a couple minor changes so we can run the tests for this, and a question on the user experience for the refresh func for manually approved features (e.g. Microsoft.HardwareSecurityModule/AzureDedicatedHSM)

Yes, I added that at line162 before refresh func, let user know manually approved features can't be managed by tf. So during refresh func, its state should not be `pending` again.


> Additionally, do the providers need to be re-registered for terraform to pick up the features that are being activated to be useful in the same apply. If we need to re-register Terraform's provider for users to have access to an activated feature, this may need integrating with azurerm_resource_provider_registration?

I don't think it needs to be re-registered. I didn't find any docs saying it's a required operation. I'll check it with service team.

ms-henglu commented:
@jackofallops , Hi, just go confirmation from service team.

It's recommended but not required, it may take up to 10 mins to get the feature registration populated to ARM cache. But in this api, it will only set state to `registered` after ARM cache updated. So we don't need to re-register rp.

katbyte reviewed (dismissed):
could we update the file & function names to reflect the new resource name? thanks

ms-henglu commented:
@katbyte , I've added a commit to rename file/func, coud you please take another look? Thanks!

jackofallops commented on website/docs/r/subscription_feature.html.markdown:
```diff
@@ -0,0 +1,51 @@
+
+# azurerm_subscription_feature
+
+Register and unregister a preview feature for the subscription. Features which `approvalType` is `AutoApproval` can be managed by terraform. 
```
Can we provide a link to a doc here to help users with what they can manage, or how they can find out if their desired feature is auto or manual approval?

jackofallops commented on azurerm/internal/services/resource/validate/feature_id_test.go:
```diff
@@ -0,0 +1,64 @@
+
+		{
+			// missing Name
+			Input: "/subscriptions/12345678-1234-9876-4563-123456789012/providers/Microsoft.Features/providers/provider1/",
```
Can we update these tests to use more accurate values?  The `providers` portion will always be a namespace dotted notation? e.g. `Microsoft.SomeAzureService` or `Foo.Bar`

jackofallops commented on website/docs/r/subscription_feature.html.markdown:
````diff
@@ -0,0 +1,51 @@
+Features can be imported using the `resource id`, e.g.
+
+```shell
+terraform import azurerm_subscription_feature.example /subscriptions/00000000-0000-0000-0000-000000000000/providers/Microsoft.Features/providers/provider1/features/Feature1
````
As with the tests, can we make this import example more representative of the real value of the feature's `providers` portion?

jackofallops reviewed (changes requested):
Hi @ms-henglu 
Apologies for the delay in looping back to this. I've left a few more comments below, after which I think this should be good to merge. 

Thanks!

ms-henglu commented on azurerm/internal/services/resource/subscription_feature_resource.go:
```diff
@@ -0,0 +1,198 @@
+		return fmt.Errorf("error unregistering feature %q: %+v", id, err)
+	}
+
+	if resp.Properties != nil && resp.Properties.State != nil {
+		if strings.EqualFold(*resp.Properties.State, Pending) {
+			return fmt.Errorf("feature (%q) which requires manual approval can not be managed by terraform", id)
+		}
+	}
```
Actually not necessary, after we register/unregister a feature, its state will become `pending` (manual approval) or `registering` (auto-approval). We won't get a `registering` at first then get a `pending` during `StateChangeConf`.

ms-henglu commented:
Hi @jackofallops ,

I've added a commit according to your suggestions. Thanks!

jackofallops reviewed (changes requested):
Hi @ms-henglu - Just a couple minor things left around updating the error message outputs after making the change to the stringer for the ID, and a tweak to the `Exists` function to account for the use of the alt subscription in the acctest environment.  After that, we should be good to merge.

Thanks!

ms-henglu commented:
@jackofallops Hi, I added a commit according to your suggestion, thanks!

tombuildsstuff commented on azurerm/internal/services/resource/subscription_feature_resource.go:
```diff
@@ -0,0 +1,197 @@
+		}
+	}
+
+	resp, err := client.Register(ctx, id.ProviderNamespace, id.Name)
+	if err != nil {
+		return fmt.Errorf("error registering feature %q: %+v", id, err)
+	}
+
+	if resp.Properties != nil && resp.Properties.State != nil {
+		if strings.EqualFold(*resp.Properties.State, Pending) {
+			return fmt.Errorf("%s which requires manual approval can not be managed by terraform", id)
+		}
+	}
```
since the Resource Provider has to be re-registered for this Feature to be usable - this shouldn't be a separate resource but must instead be embedded within [the `azurerm_resource_provider_registration` resource](https://registry.terraform.io/providers/hashicorp/azurerm/latest/docs/resources/resource_provider_registration)

tombuildsstuff commented on azurerm/internal/services/resource/subscription_feature_resource.go:
```diff
@@ -0,0 +1,197 @@
+	if !ok {
+		return fmt.Errorf("context had no deadline")
+	}
+	stateConf := &pluginsdk.StateChangeConf{
+		Pending:    []string{Unregistering},
+		Target:     []string{NotRegistered, Unregistered},
+		Refresh:    featureRegisteringStateRefreshFunc(ctx, client, *id),
+		MinTimeout: 3 * time.Minute,
+		Timeout:    time.Until(deadline),
+	}
+
+	if _, err = stateConf.WaitForStateContext(ctx); err != nil {
+		return fmt.Errorf("waiting for %s to be complete unregistering: %+v", id, err)
+	}
```
IIRC the RP also needs to be re-registered once the features disabled

tombuildsstuff reviewed (changes requested):
hi @ms-henglu 

Thanks for this PR.

Taking a look through here by design within Azure the Resource Providers need to be registered to be usable once the Feature is added/removed - as such unfortunately this doesn't make sense as a separate resource and instead will need to be integrated into [the `azurerm_resource_provider_registration` resource](https://registry.terraform.io/providers/hashicorp/azurerm/latest/docs/resources/resource_provider_registration).

As such if we can remove the separate resource and instead embed this within the `azurerm_resource_provider_registration` resource then we should be able to take another look, presumably something along the lines of:

```hcl
resource "azurerm_resource_provider_registration" "example" {
  name = "Microsoft.PolicyInsights"

  feature {
    name = "SomeFeature"
  }

  feature {
    name = "SomeOtherFeature"
  }
}
```

.. which would allow us to manage only features specified inline - and ensure we can track all enabled features within a given Resource Provider (that is, if it's Enabled it should be defined in code, if it's not it should be filtered out).

Thanks!

ms-henglu commented:
@tombuildsstuff , Hi tom, I've confirmed it with service team.
> IIRC the RP also needs to be re-registered once the features disabled

It's recommended but not required, it may take up to 10 mins to get the feature registration populated to ARM cache. But in this api, it will only set state to registered after ARM cache updated. So we don't need to re-register rp.

tombuildsstuff commented:
@ms-henglu 

> It's recommended but not required, it may take up to 10 mins to get the feature registration populated to ARM cache. But in this api, it will only set state to registered after ARM cache updated. So we don't need to re-register rp.

If it's recommended it must be for a reason, else it wouldn't be recommended all over the place 😄

Having this as a separate resource means you'd end up registering the resource provider in multiple times (which can take 5-30m depending on the RP, and would need locks within Terraform so would take longer) - as such this has to be embedded within the other resource due to that API requirement.

ms-henglu commented:
@tombuildsstuff , hi, I've embedded it into azurerm_resource_provider_registration, could you please take another look? Thanks.

grayzu commented:
Hey all, 

I have some questions/suggestions about this experience:
1. What is the expected behavior if a feature block is included? Are these the only features that should be registered?
2. What is the expected behavior if a feature block is not included? Feature is ignored or feature is unregistered?
3. How can I ensure that a feature is not registered? 

The current implementation is a bit ambiguous.

The below will clarify the expected behavior and allow more flexibility as follows:
 - block not included it is ignored (whatever the current state is not altered)
 - block included and "registered = true" registers that feature
 - block included and "registered = false" unregisters that feature
 - block removed unregisters the feature on next apply and then ignored

```terraform
resource "azurerm_resource_provider_registration" "example" {
  name = "Microsoft.PolicyInsights"

  // name and registered properties are required
  feature {
    name = "SomeFeature"
    registered = true
  }

  feature {
    name = "SomeOtherFeature"
    registered = false
  }
}
```

I know this has already gone through one rewrite but want to ensure that the experience makes sense before we release it.

ms-henglu commented:
@tombuildsstuff , Hi tom, I've added some commits according to Mark's suggestions, could you please take another look?

tombuildsstuff commented:
@grayzu 

> I have some questions/suggestions about this experience:
>
> 1. What is the expected behavior if a feature block is included? Are these the only features that should be registered?
> 2. What is the expected behavior if a feature block is not included? Feature is ignored or feature is unregistered?
> 3. How can I ensure that a feature is not registered?
>
> The current implementation is a bit ambiguous.

In my original comment I was assuming the following:

1. Features which are defined in the config should be registered (and thus, any features which are registered in the RP which aren't defined in the config should be unregistered)
2. Features could be ignored using `ignore_changes`, which since the name would be the only item means the name could be the hash value to ignore (e.g. `ignore_changes = [features.SomeFeature]`)

That said adding the `registered` flag above seems sensible to me - but we'll need a yellow box saying that features which aren't defined are ignored in the documentation - since this differs from other resources (and has to because features can be added in the background by teams, which isn't ideal but I think covers this)?

tombuildsstuff reviewed (changes requested):
👋 

I've spent the day looking into this PR and it appears there's a number of changes needed for this to work.

Whilst I've left some comments inline ([and have fixed those in this WIP commit](https://github.com/ms-henglu/terraform-provider-azurerm/compare/ticket-1851835-add-resource-feature...terraform-providers:ticket-1851835-add-resource-feature)) - having spent some time digging into this, there's an issue here at import-time which means this approach won't work and we'll need to look at something else.

Since the Feature flags are tri-state enums (that is, `NotRegistered`, `Registered` and `Unregistered`), perhaps we can work with that instead - so that if a feature isn't `NotRegistered` (e.g. it's either explicitly Registered or Unregistered) then it should be defined in the `feature` block, which gives us the information we need at import time. That's not ideal but seems to be the closest thing we can do if we want to support both the Registered and Unregistered use-case here.

WDYT @grayzu?

Thanks!

grayzu commented:
@tombuildsstuff, 

Good catch. From the swagger, these are the valid states for feature: "NotSpecified", "NotRegistered", "Pending", "Registering", "Registered", "Unregistering", "Unregistered". Based on this I would suggest that if the state is "Registering" or "Registered" at import time, we treat this as `registered = true`. If the state is "Unregistering" or "Unregistered" during import we treat this as `registered = false`.  All other states we do not import the feature.

I think this mostly aligns with your suggestion, yes?

ms-henglu commented:
@tombuildsstuff @grayzu , Hi, I just confirmed with service team. 

> NotRegistered means it has never been registered 
> Unregistered means it has been registered but unregistered later

So we can take them as the same thing.

tombuildsstuff commented:
@grayzu 

> Good catch. From the swagger, these are the valid states for feature: "NotSpecified", "NotRegistered", "Pending", "Registering", "Registered", "Unregistering", "Unregistered". Based on this I would suggest that if the state is "Registering" or "Registered" at import time, we treat this as `registered = true`. If the state is "Unregistering" or "Unregistered" during import we treat this as `registered = false`.  All other states we do not import the feature.

Yeah that sounds sensible to me, if we apply that logic to the Read function that should solve the import issue (and make this consistent across both of them).

@ms-henglu

>> NotRegistered means it has never been registered
>> Unregistered means it has been registered but unregistered later
> So we can take them as the same thing.

Nope - as per @grayzu's suggestion above, `NotRegistered` means we shouldn't track this, so:

* if this is `Registering/Registered` we can track it as `registered=true` (because the user has explicitly enabled this feature)
* if this is `Unregistered/Unregistering` we can track it as `registered=false` (because the users explicitly disabled this feature (or has enabled it previously and disabled it)
* other cases (e.g. Pending/NotRegistered/NotSpecified) we can filter out because the user doesn't want to manage this feature

As per [the changes in this commit](https://github.com/ms-henglu/terraform-provider-azurerm/compare/ticket-1851835-add-resource-feature...terraform-providers:ticket-1851835-add-resource-feature) this will need to become a Set instead of a List however, since it's possible for there to be multiple features enabled.

FWIW I could see it being useful to have an `azurerm_features` data source to go with this, perhaps with two outputs of `registered_features` and `unregistered_features` (which'd allow tooling [Sentinel/OpenPolicyAgent] to assert on those) - but I think that'd be a separate PR in it's own right.

ms-henglu commented:
@tombuildsstuff , Hi, I've updated the commit according to the latest design.

And there'll be plan changes for the old users, because feature which is not `NotRegistered` must be declared, but they can use `ignore_change` to skip them.

grayzu commented:
@tombuildsstuff, is this good to go for the next release or are we waiting on anything?

ms-henglu commented:
@tombuildsstuff , @grayzu 

Hi, I have a concern about this new design. If there's feature which is not `Auto-Approval` but registered, according to the design, user must declare this feature (registered = true), but tf can't manage feature which is not `Auto-Approval` because tf can't unregister this feature.

And I noticed there's a new version for feature API which contains the information that whether a feature is `Auto-Approval`. I think this can solve above problem. However this API doesn't have swagger and sdk.

What do you think?

ms-henglu commented:
I think maybe we can ignore the undeclared feature and ignore features during import. Or maybe we can use its v2 API, which contains the information that whether a feature is Auto-Approval, to implement it and support import as Tom suggested.

ms-henglu commented:
@grayzu @tombuildsstuff , Hi, what do you think? ⬆️

ms-henglu commented:
> @tombuildsstuff , @grayzu
> 
> Hi, I have a concern about this new design. If there's feature which is not `Auto-Approval` but registered, according to the design, user must declare this feature (registered = true), but tf can't manage feature which is not `Auto-Approval` because tf can't unregister this feature.
> 
> And I noticed there's a new version for feature API which contains the information that whether a feature is `Auto-Approval`. I think this can solve above problem. However this API doesn't have swagger and sdk.
> 
> What do you think?

@grayzu @tombuildsstuff , Hi, What do you think? ⬆️

tombuildsstuff reviewed (approved):
hi @ms-henglu 

Thanks for pushing those changes - apologies for the delayed re-review here.

I've spent some time testing this this morning and besides some minor changes to the docs to explain the updated behaviour (which I've pushed a commit for) - this otherwise LGTM 👍 

Thanks!

ms-henglu commented:
kindly ping.

tombuildsstuff commented:
Tests pass:

![Screenshot 2021-11-17 at 16 58 06](https://user-images.githubusercontent.com/666005/142235266-250e6c7e-8c16-48fa-b462-8efb9ba0e5f3.png)
